### PR TITLE
#1216 #1955 #1956 Overloaded `AddOcelot` method to support merging of configuration files to memory

### LINKED ISSUE #1216: Use merged `ocelot.json` directly from memory instead of from file
## New Feature
The `ConfigurationBuilderExtensions.AddOcelot` method merges the **ocelot.*.json** files into one **ocelot.json** file that is written back to disk and then added to the `IConfigurationBuilder`. 

Would it be possible to make another `AddOcelot` method that adds the merged JSON directly from memory into the `IConfigurationBuilder` using `builder.AddJsonStream` instead?

## Motivation for New Feature
From time to time we experience a problem at startup where something prevents the server process from writing the merged **ocelot.json** back to disk. When this happens we get an `UnauthorizedAccessException` (example below) at startup and the process must be killed. 

We have not been able to see a consistent pattern in the behavior, and we don't have access to the server since IT operation is outsourced, so this is very hard to debug.

## Exception System.UnauthorizedAccessException 
Access to the path `C:\Octopus\Applications\Dev\Xxxxxxxx\20200429.1_5\ocelot.json` is denied.
```
   at System.IO.File.WriteAllText(String path, String contents)
   at Ocelot.DependencyInjection.ConfigurationBuilderExtensions.AddOcelot(IConfigurationBuilder builder, String folder, IWebHostEnvironment env)
   at Xxxxxxxx.Program.<>c.<CreateHostBuilder>b__1_2(WebHostBuilderContext hostingContext, IConfigurationBuilder config)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass8_0.<ConfigureAppConfiguration>b__0(HostBuilderContext context, IConfigurationBuilder builder)
   at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()`
```

### PR DESCRIPTION
**Pull request by ebjornset**
## Closes #1216 #1955 
Implementation of feature suggested in issue:
- #1216 
- #1955 

## Discussion
- #1956 

## Docs
- [Framework design guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/) &rarr; [Member Overloading](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/member-overloading)

### CODE REVIEW

**Comment by ebjornset** [comment]
Hi. I just realized that I've not linked this PR correctly to issue #1216, but it seems that I don't have permissions to fix it?

**Comment by raman-m** [comment]
@ebjornset 
Hi Eirik!
Thanks for the great PR!

Unfortunately It is hard to resolve merge conflicts using GitHub UI.
Could you resolve merge conflicts please?
You need to merge **develop** branch to feature branch with rebasing. See below! 👇 

And I don't see **develop** branch in your fork!
Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, new **develop** branch will occur with all top commits!

Let me know upgrading results please!

**Comment by raman-m** [comment]
Eirik,
I thought you need some help in merge conflicts resolving.
There are no conflicts now.
The feature branch has been rebased onto [ThreeMammals:develop](../tree/develop).

And we can start code review...

**Comment by raman-m** [comment]
@pablo0219
Could you review the code please?
Your opinion is prioritized because of [past discussions](../issues/1216#issuecomment-642107288).

**Review by RaynaldM — APPROVED** [review]

File: docs/features/configuration.rst [hunk 1]
```diff
@@ -132,6 +132,16 @@ You can also give Ocelot a specific path to look in for the configuration files
...
 
+Merging configuration files to memory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default Ocelot writes the merged configuration back to disk as ``ocelot.json`` and add that file to config.
+If your server don't have write permissions to your configuration folder, you can instruct Ocelot to use the merged configuration directly from memory instead.
+
+.. code-block:: csharp
+
+    config.AddOcelot(hostingContext.HostingEnvironment, MergeOcelotJson.ToMemory)
```
Comment (raman-m) [inline docs/features/configuration.rst:143]: It will be better to show complete list of available options of the `MergeOcelotJson` enumeration, as a code snippet!

File: docs/features/configuration.rst [hunk 2]
```diff
@@ -132,6 +132,16 @@ You can also give Ocelot a specific path to look in for the configuration files
 
 Ocelot needs the HostingEnvironment so it knows to exclude anything environment specific from the algorithm. 
 
+Merging configuration files to memory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default Ocelot writes the merged configuration back to disk as ``ocelot.json`` and add that file to config.
```
Comment (raman-m) [inline docs/features/configuration.rst:138]: We need to show the default option of `MergeOcelotJson` enumeration, and explain to developers that default option is reused implicitly.

File: docs/features/configuration.rst [hunk 3]
```diff
@@ -132,6 +132,16 @@ You can also give Ocelot a specific path to look in for the configuration files
 
 Ocelot needs the HostingEnvironment so it knows to exclude anything environment specific from the algorithm. 
 
+Merging configuration files to memory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default Ocelot writes the merged configuration back to disk as ``ocelot.json`` and add that file to config.
+If your server don't have write permissions to your configuration folder, you can instruct Ocelot to use the merged configuration directly from memory instead.
```
Comment (raman-m) [inline docs/features/configuration.rst:139]: We need to explain and uncover these user scenarios!
We need to show that merging config to a memory is required mostly in all virtual environments like cloud providers (AWS, Azure, GCP).

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs [hunk 4]
```diff
@@ -39,8 +40,35 @@ public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder
 
         public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder, string folder, IWebHostEnvironment env)
         {
-            const string primaryConfigFile = "ocelot.json";
+            return builder.AddOcelot(folder, env, MergeOcelotJson.ToFile);
+        }
+
+        public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder, IWebHostEnvironment env, MergeOcelotJson mergeOcelotJson)
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs:46]: No XML docs for public extension method!

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs [hunk 5]
```diff
@@ -39,8 +40,35 @@ public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder
...
-            const string primaryConfigFile = "ocelot.json";
+            return builder.AddOcelot(folder, env, MergeOcelotJson.ToFile);
+        }
+
+        public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder, IWebHostEnvironment env, MergeOcelotJson mergeOcelotJson)
+        {
+            return builder.AddOcelot(".", env, mergeOcelotJson);
+        }
 
+        public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder, string folder, IWebHostEnvironment env, MergeOcelotJson mergeOcelotJson)
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs:51]: No XML docs for public extension method!

File: src/Ocelot/DependencyInjection/MergeOcelotJson.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Ocelot.DependencyInjection;
+
+public enum MergeOcelotJson
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/MergeOcelotJson.cs:3]: No XML docs for public enumeration!
Each option should be explained to a developer.

File: test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            },
+        };
+    }
+
+    [Fact]
+    public void Should_run_with_config_merged_to_memory()
+    {
+        this.Given(x => _steps.GivenThereIsAConfiguration(_globalConfig, TestConfiguration.ConfigurationPartPath("global")))
+            .When(x => _steps.WhenOcelotIsRunningMergedConfig(MergeOcelotJson.ToMemory))
+            .Then(x => TheOcelotJsonFileExists(false))
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs:46]: I suggest to show that merged config contains some important option by ensuring it exists in merged object,
It can be additional check like `.And(x => TheOptionPropertyExistsInMergedObject())`

File: test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            .BDDfy();
+    }
+
+    [Fact]
+    public void Should_run_with_config_merged_to_file()
+    {
+        this.Given(x => _steps.GivenThereIsAConfiguration(_globalConfig))
+            .When(x => _steps.WhenOcelotIsRunningMergedConfig(MergeOcelotJson.ToFile))
+            .Then(x => TheOcelotJsonFileExists(true))
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs:55]: Consider extra check by `.And(x => TheOptionPropertyExistsInMergedObject())`

File: test/Ocelot.AcceptanceTests/Steps.cs [hunk 9]
```diff
@@ -214,30 +214,12 @@ public void ThenTheResponseBodyHeaderIs(string key, string value)
...
-                });
-
-            _ocelotServer = new TestServer(_webHostBuilder);
+            StartMinimalTestServerWithOcelotConfiguration((hostingContext, config) => config.AddJsonFile("ocelot.json", optional: false, reloadOnChange: shouldReload));
+        }
 
-            _ocelotClient = _ocelotServer.CreateClient();
+        public void WhenOcelotIsRunningMergedConfig(MergeOcelotJson mergeOcelotJson)
+        {
+            StartMinimalTestServerWithOcelotConfiguration((hostingContext, config) => config.AddOcelot(hostingContext.HostingEnvironment, mergeOcelotJson));
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/Steps.cs:222]: How many references does this method have?

If it has one reference only, consider moving the helper to the `ConfigurationMergeTests` class as public static definition plz.

File: test/Ocelot.UnitTests/DependencyInjection/ConfigurationBuilderExtensionsTests.cs [hunk 10]
```diff
@@ -36,7 +36,7 @@ public ConfigurationBuilderExtensionsTests()
             _hostingEnvironment = new Mock<IWebHostEnvironment>();
 
             // Clean up config files before each test
-            var subConfigFiles = new DirectoryInfo(".").GetFiles("ocelot.*.json");
+            var subConfigFiles = new DirectoryInfo(".").GetFiles("ocelot*.json");
```
Comment (raman-m) [inline test/Ocelot.UnitTests/DependencyInjection/ConfigurationBuilderExtensionsTests.cs:39]: Why did you remove the dot (`.` sign) from file template?
How will it reflect to current docs? See: [Merging configuration files](../../blob/develop/docs/features/configuration.rst#merging-configuration-files)

**Review by raman-m — CHANGES_REQUESTED** [review]
I've pointed my attention to these new extensions because they are highly reused by developers. These extension-methods are critical in Ocelot core.
And, it makes sense to develop them a bit more.

Could you fix the issues please?

**Comment by raman-m** [comment]
@ebjornset Hi Eirik!
I've rebased the feature branch. In current release the `AddOcelot` extension methods were changed. Also we've refactored acceptance tests and especially `Steps` class.
I tried to merge develop and rebase the feature branch in the best logical manner. But now some unit tests are failing.
**Could you look into these failed tests please?**

File: test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs (hunk 7 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1227/commits/ce5e2aeb994d12fe3681391f92e3cf9d77733127 ✔️

File: test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs (hunk 8 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/ConfigurationMergeTests.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1227/commits/ce5e2aeb994d12fe3681391f92e3cf9d77733127 ✔️

File: test/Ocelot.AcceptanceTests/Steps.cs (hunk 9 above)
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/Steps.cs]: not an issue!

File: test/Ocelot.UnitTests/DependencyInjection/ConfigurationBuilderExtensionsTests.cs (hunk 10 above)
Comment (raman-m) [reply test/Ocelot.UnitTests/DependencyInjection/ConfigurationBuilderExtensionsTests.cs:27]: not an issue!
The code removes all files of the old/previous test.

**Review by raman-m — APPROVED** [review]
### TODO
- [ ] Update [dependencyinjection.rst](https://github.com/ebjornset/Ocelot/blob/feat/1216-Merge-ocelot-json-to-memory/docs/features/dependencyinjection.rst) with new methods
- [ ] Check references for [dependencyinjection.rst](https://github.com/ebjornset/Ocelot/blob/feat/1216-Merge-ocelot-json-to-memory/docs/features/dependencyinjection.rst) aka `ref:`

**Review by ggnaegi — APPROVED** [review]
Ok, thanks.

**Comment by raman-m** [comment]
Ready for delivery! ✅ 
- Code review ✔️✔️
- Team approvals ✔️
- Unit testing ✔️ 
- Acceptance testing ✔️ 
- Reviewed and updated docs ✔️✔️
